pinot-segment-local/src/test/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/NativeMutableTextIndexConcurrentTest.java

```diff
@@ -0,0 +1,251 @@
+import static org.testng.Assert.assertTrue;
+
+
+public class NativeMutableTextIndexConcurrentTest {
+  private ExecutorService _threadPool;
+  private Set<String> _resultSet;
+
+  @BeforeClass
+  private void setup() {
+    _threadPool = Executors.newFixedThreadPool(10);
+    _resultSet = new ConcurrentSkipListSet<>();
+  }
+
+  @AfterClass
+  private void shutDown() {
+    _threadPool.shutdown();
+  }
+
+  @Test
+  public void testConcurrentWriteAndRead()
+      throws InterruptedException, IOException {
+    CountDownLatch countDownLatch = new CountDownLatch(2);
+    List<String> words = new ArrayList<>();
+    words.add("ab");
+    words.add("abba");
+    words.add("aba");
+    words.add("bab");
+    words.add("cdd");
+    words.add("efg");
+
+    try (NativeMutableTextIndex textIndex = new NativeMutableTextIndex("testFSTColumn")) {
+      _threadPool.submit(() -> {
+        try {
+          performReads(textIndex, words, 20, 200, countDownLatch);
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        }
+      });
+
+      _threadPool.submit(() -> {
+        try {
+          performWrites(textIndex, words, 10, countDownLatch);
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        }
+      });
+
+      countDownLatch.await();
+    }
+
+    assertEquals(_resultSet.size(), words.size());
+
+    assertTrue(_resultSet.contains("ab"), "ab not found in result set");
+    assertTrue(_resultSet.contains("abba"), "abba not found in result set");
+    assertTrue(_resultSet.contains("aba"), "aba not found in result set");
+    assertTrue(_resultSet.contains("bab"), "bab not found in result set");
+    assertTrue(_resultSet.contains("cdd"), "cdd not found in result set");
+    assertTrue(_resultSet.contains("efg"), "efg not found in result set");
+  }
+
+  @Test
+  public void testConcurrentWriteWithMultipleThreads()
+      throws InterruptedException, IOException {
+    List<String> firstThreadWords = new ArrayList<>();
+    List<String> secondThreadWords = new ArrayList<>();
+    List<String> mergedThreadWords = new ArrayList<>();
+    CountDownLatch countDownLatch = new CountDownLatch(3);
+    firstThreadWords.add("ab");
+    firstThreadWords.add("abba");
+    firstThreadWords.add("aba");
+    secondThreadWords.add("bab");
+    secondThreadWords.add("cdd");
+    secondThreadWords.add("efg");
+
+    mergedThreadWords.addAll(firstThreadWords);
+    mergedThreadWords.addAll(secondThreadWords);
+
+    try (NativeMutableTextIndex textIndex = new NativeMutableTextIndex("testFSTColumn")) {
+      _threadPool.submit(() -> {
+        try {
+          performReads(textIndex, mergedThreadWords, 20, 200, countDownLatch);
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        }
+      });
+
+      _threadPool.submit(() -> {
+        try {
+          performWrites(textIndex, firstThreadWords, 10, countDownLatch);
```

Comment:
We don't need to test multiple writer scenario (although it seems work). For mutable index, pinot always follow single writer multiple readers. This gives us a lot of optimization opportunities since we don't need to worry about concurrent modification of the index

Reply:
I am leaving the test as is -- if nothing else, it demonstrates a capability that is present should we need it someday